llvm/test/MC/X86/apx/rex2-bit-att.s

```diff
@@ -0,0 +1,240 @@
+         leal	(%rax), %r28d
+
+# CHECK: leal	(%rax), %r29d
+# CHECK: encoding: [0xd5,0x44,0x8d,0x28]
+         leal	(%rax), %r29d
+
+# CHECK: leal	(%rax), %r30d
+# CHECK: encoding: [0xd5,0x44,0x8d,0x30]
+         leal	(%rax), %r30d
+
+# CHECK: leal	(%rax), %r31d
+# CHECK: encoding: [0xd5,0x44,0x8d,0x38]
+         leal	(%rax), %r31d
+
+## X bit
+
+# CHECK: leal	(,%r16), %eax
+# CHECK: encoding: [0xd5,0x20,0x8d,0x04,0x05,0x00,0x00,0x00,0x00]
+         leal	(,%r16), %eax
+
+# CHECK: leal	(,%r17), %eax
+# CHECK: encoding: [0xd5,0x20,0x8d,0x04,0x0d,0x00,0x00,0x00,0x00]
+         leal	(,%r17), %eax
+
+# CHECK: leal	(,%r18), %eax
+# CHECK: encoding: [0xd5,0x20,0x8d,0x04,0x15,0x00,0x00,0x00,0x00]
+         leal	(,%r18), %eax
+
+# CHECK: leal	(,%r19), %eax
+# CHECK: encoding: [0xd5,0x20,0x8d,0x04,0x1d,0x00,0x00,0x00,0x00]
+         leal	(,%r19), %eax
+
+# CHECK: leal	(,%r20), %eax
+# CHECK: encoding: [0xd5,0x20,0x8d,0x04,0x25,0x00,0x00,0x00,0x00]
+         leal	(,%r20), %eax
+
+# CHECK: leal	(,%r21), %eax
+# CHECK: encoding: [0xd5,0x20,0x8d,0x04,0x2d,0x00,0x00,0x00,0x00]
+         leal	(,%r21), %eax
+
+# CHECK: leal	(,%r22), %eax
+# CHECK: encoding: [0xd5,0x20,0x8d,0x04,0x35,0x00,0x00,0x00,0x00]
+         leal	(,%r22), %eax
+
+# CHECK: leal	(,%r23), %eax
+# CHECK: encoding: [0xd5,0x20,0x8d,0x04,0x3d,0x00,0x00,0x00,0x00]
+         leal	(,%r23), %eax
+
+# CHECK: leal	(,%r24), %eax
+# CHECK: encoding: [0xd5,0x22,0x8d,0x04,0x05,0x00,0x00,0x00,0x00]
+         leal	(,%r24), %eax
+
+# CHECK: leal	(,%r25), %eax
+# CHECK: encoding: [0xd5,0x22,0x8d,0x04,0x0d,0x00,0x00,0x00,0x00]
+         leal	(,%r25), %eax
+
+# CHECK: leal	(,%r26), %eax
+# CHECK: encoding: [0xd5,0x22,0x8d,0x04,0x15,0x00,0x00,0x00,0x00]
+         leal	(,%r26), %eax
+
+# CHECK: leal	(,%r27), %eax
+# CHECK: encoding: [0xd5,0x22,0x8d,0x04,0x1d,0x00,0x00,0x00,0x00]
+         leal	(,%r27), %eax
+
+# CHECK: leal	(,%r28), %eax
+# CHECK: encoding: [0xd5,0x22,0x8d,0x04,0x25,0x00,0x00,0x00,0x00]
+         leal	(,%r28), %eax
+
+# CHECK: leal	(,%r29), %eax
+# CHECK: encoding: [0xd5,0x22,0x8d,0x04,0x2d,0x00,0x00,0x00,0x00]
+         leal	(,%r29), %eax
+
+# CHECK: leal	(,%r30), %eax
+# CHECK: encoding: [0xd5,0x22,0x8d,0x04,0x35,0x00,0x00,0x00,0x00]
+         leal	(,%r30), %eax
+
+# CHECK: leal	(,%r31), %eax
+# CHECK: encoding: [0xd5,0x22,0x8d,0x04,0x3d,0x00,0x00,0x00,0x00]
+         leal	(,%r31), %eax
+
+## B bit
+
+# CHECK: leal	(%r16), %eax
+# CHECK: encoding: [0xd5,0x10,0x8d,0x00]
+         leal	(%r16), %eax
+
+# CHECK: leal	(%r17), %eax
+# CHECK: encoding: [0xd5,0x10,0x8d,0x01]
+         leal	(%r17), %eax
+
+# CHECK: leal	(%r18), %eax
+# CHECK: encoding: [0xd5,0x10,0x8d,0x02]
+         leal	(%r18), %eax
+
+# CHECK: leal	(%r19), %eax
+# CHECK: encoding: [0xd5,0x10,0x8d,0x03]
+         leal	(%r19), %eax
+
+# CHECK: leal	(%r20), %eax
+# CHECK: encoding: [0xd5,0x10,0x8d,0x04,0x24]
+         leal	(%r20), %eax
+
+# CHECK: leal	(%r21), %eax
+# CHECK: encoding: [0xd5,0x10,0x8d,0x45,0x00]
+         leal	(%r21), %eax
```

Comment:
Why the instrution length changed?

Reply:
It's explained at X86MCCodeEmitter.cpp:814.
```
             // Base reg can't be EBP/RBP/R13/R21/R29 as that would end up with
             // '5' as the base field, but that is the magic [*] nomenclature
             // that indicates no base when mod=0. For these cases we'll emit a
             // 0 displacement instead.
```

Reply:
Should the disassembly be `leal (, %r21), %eax` in this case no matter what the assembly is?
And why `leal (%r20), %eax` has 5 bytes too?

Reply:
No. `leal (, %r21), %eax` has a different encoding from `leal (, %r21), %eax`.  If it's a base reg, then we decode it to a base reg.

> Why leal (%r20), %eax has 5 bytes too?

It's explained at X86MCCodeEmitter.cpp:763
```
    // If the base is not EBP/ESP/R12/R13/R20/R21/R28/R29 and there is no
    // displacement, use simple indirect register encoding, this handles
    // addresses like [EAX]. The encoding for [EBP], [R13], [R20], [R21], [R28]
    // or [R29] with no displacement means [disp32] so we handle it by emitting
    // a displacement of 0 later.
```

Reply:
The comment doesn't match the code here.
[R21] must be encoded with a displacement 0, but [R20] doesn't. It's just encoded with a SIB byte.

Reply:
You are right. Updated the comment at X86MCCodeEmitter.cpp:756
```
      // The SIB byte must be used if the base is ESP/RSP/R12/R20/R28, all of
      // which encode to an R/M value of 4, which indicates that a SIB byte is
      // present.
```